dali/operators/audio/nonsilence_op_gpu.cu

```diff
@@ -0,0 +1,216 @@
+          NonsilenceOperator<GPUBackend>(spec) {}
+
+  ~NonsilenceOperatorGpu() override = default;
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperatorGpu);
+
+ protected:
+  void RunImpl(workspace_t<GPUBackend> &ws) override {
+    auto dtype = ws.template Input<GPUBackend>(0).type();
+    TYPE_SWITCH(dtype, type2id, T, NONSILENCE_TYPES, (
+      RunImplTyped<T>(ws);
+    ), DALI_FAIL(make_string("Unsupported input type: ", dtype));)  // NOLINT
+  }
+
+ private:
+  kernels::MaxGPU<float, float> max_kernel_;
+  using Predicate = NotLess<float>;
+  kernels::FindRegionGPU<float, Predicate> find_region_;
+
+  /**
+   * @brief Computes mean moving mean squre of an audio signal
+   */
+  template <typename T>
+  void CalcMMS(kernels::KernelContext &ctx, TensorListView<StorageGPU, float, 1> &mms,
+               const TensorListView<StorageGPU, const T, 1> &in) {
+    kernels::signal::MovingMeanSquareGpu<T> kernel;
+    kernels::signal::MovingMeanSquareArgs args{window_length_, -1};
+    kernel.Run(ctx, mms, in, args);
+  }
+
+  /**
+   * @brief Calculates the maximum value of an input
+   */
+  void CalcMax(kernels::KernelContext &ctx, TensorListView<StorageGPU, float, 0> &max,
+               const TensorListView<StorageGPU, float, 1> &in) {
+    std::array<int, 1> axes = { 0 };
+    max_kernel_.Setup(ctx, in.shape, make_cspan(axes), false, false);
+    max_kernel_.Run(ctx, max, in);
+  }
+
+  /**
+   * @brief Calculates the beginning and length of the nonsilent region
+   */
+  void CalcNonsilentRegion(kernels::KernelContext &ctx,
+                           TensorListView<StorageGPU, int32_t, 0> &begin,
+                           TensorListView<StorageGPU, int32_t, 0> &len,
+                           TensorListView<StorageGPU, float, 1> &mms) {
+    int nsamples = mms.num_samples();
+    int G = div_ceil(nsamples, 256), B = 256;  // for the simpler kernels
+
+    TensorListShape<0> scalar_sh(nsamples);
+    auto predicates =
+        ctx.scratchpad->AllocTensorList<mm::memory_kind::device, NotLess<float>>(scalar_sh);
+
+    if (!reference_max_) {
+      auto predicates_cpu =
+        ctx.scratchpad->AllocTensorList<mm::memory_kind::pinned, NotLess<float>>(scalar_sh);
+      for (int i = 0; i < nsamples; i++) {
+        kernels::signal::DecibelToMagnitude<float> db2mag(10.f, reference_power_[i]);
+        (*predicates_cpu.data[i]) = {db2mag(cutoff_db_[i])};
+      }
+      assert(predicates_cpu.is_contiguous());
+      assert(predicates.is_contiguous());
+      CUDA_CALL(cudaMemcpyAsync(predicates.data[0], predicates_cpu.data[0],
+                                sizeof(Predicate) * nsamples, cudaMemcpyHostToDevice,
+                                ctx.gpu.stream));
+    } else {
+      // Predicates need to initialized on the device, because reference power is on the GPU
+      // (needs to be calculated as a pre-step)
+      auto max_mms = ctx.scratchpad->AllocTensorList<mm::memory_kind::device, float>(scalar_sh);
+      CalcMax(ctx, max_mms, mms);
```

Comment:
The memory used by CalcMMS could be freed - reusing the same dynamic scratchpad will increase the memory consumption.